src/System.Data.Common/src/System/Data/DataView.cs

```diff
@@ -1330,9 +1330,9 @@ internal void MaintainDataView(ListChangedType changedType, DataRow row, bool tr
                         _addNewRow = null;
                         _addNewMoved = new ListChangedEventArgs(ListChangedType.ItemMoved, index, Count - 1);
                     }
-                    else if (!_rowViewCache.ContainsKey(row))
+                    else if (_rowViewCache.TryAdd(row, buffer ?? new DataRowView(this, row)))
```

Comment:
This should be reverted: if buffer is null, we'll end up allocating a new DataRowView even if row is already in the _rowViewCache.